pep-0xxx.rst

```diff
@@ -0,0 +1,797 @@
+            then read into the buffer instead, and return the number of bytes
+            read.
+
+            Raise ``WantReadError`` or ``WantWriteError`` if there is
+            insufficient data in either the input or output buffer and the
+            operation would have caused data to be written or read.
+
+            As at any time a re-negotiation is possible, a call to ``read()``
+            can also cause write operations.
+            """
+
+        @abstractmethod
+        def write(self, buf: Any) -> int:
+            """
+            Write ``buf`` in encrypted form to the output buffer and return the
+            number of bytes written. The ``buf`` argument must be an object
+            supporting the buffer interface.
+
+            Raise ``WantReadError`` or ``WantWriteError`` if there is
+            insufficient data in either the input or output buffer and the
+            operation would have caused data to be written or read.
+
+            As at any time a re-negotiation is possible, a call to ``write()``
+            can also cause read operations.
+            """
+
+        @abstractmethod
+        def do_handshake(self) -> None:
+            """
+            Performs the TLS handshake. Also performs certificate validation
+            and hostname verification.
+            """
+
+        @abstractmethod
+        def cipher(self) -> Optional[Cipher]:
+            """
+            Returns the Cipher entry for the cipher that has been negotiated on
+            the connection. If no connection has been negotiated, returns
+            ``None``.
+            """
+
+        @abstractmethod
+        def negotiated_protocol(self) -> Optional[NextProtocol]:
+            """
+            Returns the protocol that was selected during the TLS handshake.
+            This selection may have been made using ALPN, NPN, or some future
+            negotiation mechanism.
+
+            If ``Context.set_inner_protocols()`` was not called, if the other
+            party does not support protocol negotiation, if this socket does
+            not support any of the peer's proposed protocols, or if the
+            handshake has not happened yet, ``None`` is returned.
+            """
+
+        @property
+        @abstractmethod
+        def context(self) -> Context:
+            """
+            The ``Context`` object this socket is tied to.
+            """
+
+        @context.setter
+        @abstractmethod
+        def context(self, value: Context) -> None:
+            """
+            Set the value of the ``Context`` object this socket is tied to.
+            This operation (changing the context) may not always be supported.
+            """
+
+        @abstractproperty
+        def negotiated_tls_version(self) -> Optional[TLSVersion]:
+            """
+            The version of TLS that has been negotiated on this connection.
+            """
+
+        @abstractmethod
+        def shutdown(self) -> None:
```

Comment:
Well I was just looking at the current ``MemoryBio`` implementation it had a property to determine if the buffer has shutdown or not. I've never actually used the memory bio so I don't know, but presumably it'd be useful to prevent getting an exception if you've already closed the buffer or as a signal for other things that this is finished. I could imagine something like:

```python

tls_buffer = ...
while not tls_buffer.is_shutdown:
    data_to_write = a_write_queue.get(block=True)
    if data_to_write is AShutdownSigil:
        tls_buffer.shutdown()
    else:
        tls_buffer.write(data_to_write)
```

There are other ways to write this of course (a ``while True`` loop with a ``break`` comes to mind) but it feels useful to me?